### PR TITLE
Introduce a base class for aws triggers

### PR DESCRIPTION
**Pull request by vandonr-amz**
There has been a lot of triggers added recently, and they all reuse more or less the same code, with some variations.
This leads to many (small) inconsistencies in trigger code, and also in a lot of duplicated code.

We can package all this in a single base class and reuse it as much as possible, which is what I'm doing here.

Adapting all triggers to this base class is also a good way to resolve some of the inconsistencies.

---

### REVIEWERS:

1. I'd advise starting by looking at the newly added file `base_trigger.py`
2. Then, look at how the other triggers were changed to use that new base trigger
3. Then, look at the small changes in the operators/sensors
4. Then, look at the tests.

cc @syedahsn @ferruzzi @vincbeck

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/athena.py
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/hooks/athena.py:256]: this is a somewhat unrelated fix that allows specifying a sleep time of 0 in unit tests. Without this, athena unit tests were taking 30s each

File: airflow/providers/amazon/aws/triggers/athena.py
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/triggers/athena.py:41]: this trigger was added in #32186 merged on June 27th, last provider release was on June 20th, so this breaking change is OK.

File: airflow/providers/amazon/aws/triggers/batch.py
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/triggers/batch.py:251]: this trigger was added in #32036 merged on June 27th, last provider release was on June 20th, so this breaking change is OK.

File: airflow/providers/amazon/aws/triggers/ecs.py
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/triggers/ecs.py:33]: this trigger was added in #31881 merged on June 23rd, last provider release was on June 20th, so this breaking change is OK.

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (vandonr-amz) [inline airflow/providers/amazon/aws/triggers/eks.py:121]: this trigger was added in #32165 merged on June 26th, last provider release was on June 20th, so this breaking change is OK.

File: tests/providers/amazon/aws/triggers/test_athena.py
Comment (vandonr-amz) [inline tests/providers/amazon/aws/triggers/test_athena.py:23]: This is probably best viewed in side-by-side diff. I removed existing tests because there is no logic anymore in individual triggers.
Instead, I'm testing the only thing that can be broken, which is the serialization/deserialization.
To do that, I do a cycle of serialize-deserialize-reserialize and I compare the serialized data. It'd probably be better to compare the instances, but at least comparing the serialized output can be done with a simple `==`

I copy-pasted the same test for all triggers inheriting from the base, because I think it's better to have it in their respective files ? It could also be a parametrized test with many cases in test_base_trigger to avoid the code duplication, open to hear opinion about it.

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (vincbeck) [inline airflow/providers/amazon/aws/triggers/base_trigger.py]: Is not `None` included in `Any`?

**Review by vincbeck — COMMENTED** [review]
A simple nit but overall I love it!

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/base_trigger.py]: hmm yeah probably. I'll remove it from here and add it to the doc so that it's more obvious that it's an accepted value.

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/amazon/provider.yaml
Comment (eladkal) [inline airflow/providers/amazon/provider.yaml]: Do we need the `_trigger` suffix?

For example we have:
`from airflow.hooks.base import BaseHook`

Comment (vandonr-amz) [reply airflow/providers/amazon/provider.yaml]: hmm, yeah, I could name the file `base` instead of `base_trigger`, but what would it change ? How is it better ?

File: airflow/providers/amazon/aws/hooks/athena.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/hooks/athena.py:256]: This is fine, but you can also just mock the sleep function no?

File: airflow/providers/amazon/provider.yaml
Comment (eladkal) [reply airflow/providers/amazon/provider.yaml]: When it comes to naming things my philosophy is the shorter - the better (as long as not compromising name being self explained)

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (syedahsn) [inline airflow/providers/amazon/aws/triggers/base_trigger.py]: This should specify which fields are serialized by default. For example, we don't want someone passing `aws_conn_id` as a serialized field, because even though it needs to be serialized, it is done by default. What do you think?

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (syedahsn) [inline airflow/providers/amazon/aws/triggers/eks.py:121]: Why are you making a separate trigger for create/delete here? Is it no longer possible to use a generic Trigger if the responsibility is to just poll for a particular state (depending on `waiter_name`)?

File: airflow/providers/amazon/aws/hooks/athena.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/athena.py:256]: I'll leave this up to you, but it seems easier to read as 

```suggestion
                waiter_delay=sleep_time if sleep_time else self.sleep_time,
```

which should also boil down further to just
 
```suggestion
                waiter_delay=sleep_time or self.sleep_time,
```
image

File: airflow/providers/amazon/aws/operators/ecs.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/operators/ecs.py:147]: I'll make the same suggestion I made in Syed's PR... feel free to ignore it, but I see a lot of repetition in the calls to the Triggers and the Waiters.  It may be convenient to fold the waiter parameters into a new self.waiter_config dict in the Operator's init and just unpack it in the Trigger and Waiter calls, something like:

```
self.waiter_config = {
    cluster_arn=cluster_details["clusterArn"],
    waiter_delay=self.waiter_delay,
    waiter_max_attempts=self.waiter_max_attempts,
    aws_conn_id=self.aws_conn_id,
    region=self.region,
}
```

then all of the waiters and triggers can reuse that cleanly like:

```trigger=ClusterActiveTrigger(**self.waiter_config)```

Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/ecs.py:147]: oorrrrrrr maybe that's another use for a BaseAwsOperator class... it could have a helper method that generates a `self.trigger_config` and/or a `self.waiter_config` in the constructor...

File: airflow/providers/amazon/aws/sensors/batch.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/sensors/batch.py:116]: Are we no longer worried about an IndexError here if 'status' is not in event??

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/base_trigger.py]: Not positive, but should this raise NotImplementedException like they did [here](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/sensors/sagemaker.py#L75) to enforce that?

File: airflow/providers/amazon/aws/triggers/ecs.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/ecs.py]: I'm all for shorter docs, but maybe "until it is `active`"?

File: airflow/providers/amazon/aws/triggers/ecs.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/ecs.py]: Possible copypasta, I think this should be "until it is inactive"??

File: airflow/providers/amazon/aws/triggers/ecs.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/ecs.py:72]: I may be over-complicating this, but it looooks like the only difference between `ClusterActiveTrigger` and `ClusterInactiveTrigger` is the `waiter_name` value and the two message.  In which case, why not have them both inherit a `ClusterStatusTrigger` which accepts those three values and drop all the repetition?


Same below where you split other Triggers out into status-specific ones.

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/eks.py:48]:
```suggestion
        if region:
```

File: airflow/providers/amazon/aws/triggers/emr.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/emr.py:124]: Here and elsewhere:

```suggestion
        if poll_interval or max_attempts:
```

File: airflow/providers/amazon/aws/triggers/emr.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/triggers/emr.py:174]:
```suggestion
        if poll_interval or max_attempts:
```

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/base_trigger.py]: hmm, yeah. From the way it's used, it'd be mostly harmless, just unnecessary. I can add it to the comments.

File: tests/providers/amazon/aws/triggers/test_base_trigger.py
Comment (ferruzzi) [inline tests/providers/amazon/aws/triggers/test_base_trigger.py]:
```suggestion
    """An empty implementation that allows instantiation for tests."""
```


Would fail inclusive language check.

File: tests/providers/amazon/aws/triggers/test_batch.py
Comment (ferruzzi) [inline tests/providers/amazon/aws/triggers/test_batch.py:35]: :chefkiss:

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/eks.py:121]: I remember you said back in the day that you preferred separate triggers so that the status/failure messages can be more descriptive ;)

Also, given the lower footprint of triggers created this way, I think it's ok to have specific triggers for each thing.

File: airflow/providers/amazon/aws/hooks/athena.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/hooks/athena.py:256]: I think I wrote it this way to allow `sleep_time = 0`. `is None` will keep the zero value, but `or` will replace it with the alternative, and being able to specify a sleep time of 0 is useful in tests.

Comment (vandonr-amz) [reply airflow/providers/amazon/aws/hooks/athena.py:256]: didn't I put a self-review comment about that ? 🤔

File: tests/providers/amazon/aws/triggers/test_eks.py
Comment (ferruzzi) [inline tests/providers/amazon/aws/triggers/test_eks.py:38]: I do love a well-parameterized test.  It's beautiful.

File: airflow/providers/amazon/aws/hooks/athena.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/hooks/athena.py:256]: I think just setting sleep_time=0 is sooo much simpler & cleaner & easier to read

File: airflow/providers/amazon/aws/operators/ecs.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/operators/ecs.py:147]: yeah, maybe something to do in a base operator, because there is usually only one trigger being built per operator, so you'd have to pack those parameters once per operator, and use them once, and it's not really saving anything ? Just making the code harder to read by moving the parameters away from where they are passed ?

File: airflow/providers/amazon/aws/sensors/batch.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/sensors/batch.py:116]: no I don't think so. The event is something we control and that we send to ourselves. If it doesn't contain a "status" key, it means something bad happened, either the event is not formatted properly, or an event was routed to the wrong place, and we definitely want the error to blow up.

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/base_trigger.py]: the `@abstractmethod` annotation is taking care of that (and more)

File: airflow/providers/amazon/aws/operators/ecs.py
Comment (ferruzzi) [reply airflow/providers/amazon/aws/operators/ecs.py:147]: I'll buy that.  We should keep a list somewhere of all these little "would be nice in a BaseAwsOperator" things... I can never think of one big convincing reason to put in the effort, but there have been a few smaller conveniences here and there that I don't remember.

File: airflow/providers/amazon/aws/triggers/ecs.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/ecs.py:72]: the status/failure messages change as well.
But you have the same comment as syed here https://github.com/apache/airflow/pull/32274#discussion_r1253537693

File: airflow/providers/amazon/aws/triggers/base_trigger.py
Comment (ferruzzi) [reply airflow/providers/amazon/aws/triggers/base_trigger.py]: oooohhhh.  I'll need to look into that more.    Sorry about that.

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/eks.py:48]: I tend to be cautious with checks like this, because it'll trigger on empty string as well (and empty array, empty dict, etc.), and I've been bitten by this several times.
Here it's not a problem, but I prefer to be more explicit in general.

File: airflow/providers/amazon/aws/triggers/emr.py
Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/emr.py:124]: same remark as above, I think I'd like the warning to show even if the value is zero, hence the explicit check

File: airflow/providers/amazon/aws/triggers/eks.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/triggers/eks.py:121]: Ah, I ended up adopting your method haha. Alright, so just to be clear, I am going to be creating a separate Trigger for each operator again.

>Also, given the lower footprint of triggers created this way, I think it's ok to have specific triggers for each thing.

Definitely. Big +1 for reducing the repeated code.

File: airflow/providers/amazon/aws/triggers/emr.py
Comment (syedahsn) [inline airflow/providers/amazon/aws/triggers/emr.py:221]: is this supposed to be 60?

Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/emr.py:221]: it's supposed to be ∞ lol
we don't really have a clear posture on this, some triggers have sensible max attempts values, some others wait forever...
What I don't like with setting a "low" value like 60 is that if the user sets the poll interval to 1 for instance, it ends in 1 minute, which might be annoying.
Overall, I think it should be a really high value, and users should set it themselves if they care.

File: tests/providers/amazon/aws/triggers/test_base_trigger.py
Comment (syedahsn) [inline tests/providers/amazon/aws/triggers/test_base_trigger.py]: What about testing things like multiple attempts, attempts exceeded etc? I suppose those are covered in [`test_waiter_with_logging.py`](https://github.com/apache/airflow/blob/main/tests/providers/amazon/aws/utils/test_waiter_with_logging.py) so they don't need to be repeated.

**Review by syedahsn — APPROVED** [review]
Some minor comments, but overall, looks really good.

File: tests/providers/amazon/aws/triggers/test_base_trigger.py
Comment (vandonr-amz) [reply tests/providers/amazon/aws/triggers/test_base_trigger.py]: yes indeed, this code is not in the base trigger, so it doesn't have to be tested here imho

**Review by ferruzzi — APPROVED** [review]
Looks like all of my comments were addressed.

File: airflow/providers/amazon/aws/triggers/emr.py
Comment (syedahsn) [reply airflow/providers/amazon/aws/triggers/emr.py:221]: I think a good compromise is to keep it the same as the default one used by boto. That way, we don't inadvertently mess things up for users who might be relying on a Task to fail after a certain amount of time, because they didn't change the default value

Comment (vandonr-amz) [reply airflow/providers/amazon/aws/triggers/emr.py:221]: ok, but in this particular case, the existing behavior was to wait forever, so....

**Comment by vandonr-amz** [comment]
this can be merged as is only if https://github.com/apache/airflow/issues/32389 does not proceed.
If it does, I'll have to change the parts that are "breaking" to surround them with deprecation warnings.

**Comment by vincbeck** [comment]
Amazon provider package is excluded from RC2, therefore we can merge it. See #32389

**Pull request merged**
